src/crypto/OlmDevice.js

```diff
@@ -177,36 +181,30 @@ OlmDevice.getOlmVersion = function() {
 
 /**
  * extract our OlmAccount from the crypto store and call the given function
- * with the account object and a 'save' function which returns a promise.
- * The `account` will be freed as soon as `func` returns - even if func returns
- * a promise
+ * with the account object
+ * The `account` object is useable only within the callback passed to this
+ * function and will be freed as soon the callback returns. It is *not*
+ * useable for the rest of the lifetime of the transaction.
  *
+ * @param {*} txn
  * @param {function} func
- * @return {object} result of func
  * @private
  */
-OlmDevice.prototype._getAccount = async function(func) {
-    let result;
-
-    await this._cryptoStore.endToEndAccountTransaction((pickledAccount, save) => {
-        // Olm has a limited heap size so we must tightly control the number of
-        // Olm account objects in existence at any given time: once created, it
-        // must be destroyed again before we await.
+OlmDevice.prototype._getAccount = function(txn, func) {
+    this._cryptoStore.getAccount(txn, (pickledAccount) => {
         const account = new Olm.Account();
         try {
             account.unpickle(this._pickleKey, pickledAccount);
-
-            result = func(account, () => {
-                const pickledAccount = account.pickle(this._pickleKey);
-                save(pickledAccount);
-            });
+            func(account);
```

Comment:
should we not return the result here and at [line 194](https://github.com/matrix-org/matrix-js-sdk/pull/582/commits/f11a58e2cb2b98dff8f77e3031db18dc7d0bb359#diff-d0d27d71268ce9cfb61c9b3bc338fd79R194)?

Reply:
Possibly, but as per other comment it could get quite hairy, so probably simpler to just never do it at all.